Material Creation, Properties, Flags, and Modification documentation (#3912)
GregTechCEu/GregTech-Modern: docs/content/Modpacks/Materials-and-Elements/01-Material-Creation.md

Materials are in-game items or fluids. They can be dusts, ingots, gems, fluids and all their derivatives.
To make a new material, write an `event.create()` call in the registering function, like in the examples.
Write inside the parentheses the name of the material inside `''` or `""`.

> (**NOTE**: to add a material that is present on the periodic table, but doesn't have any in-game items/fluids, look below for how to do it)

Reviewer: Docs should use relative links instead of absolute ones, e.g. `[material modification page](./Modifying-Existing-Materials.md)`. This way they also work in websites that aren't the main one (like local MKDocs Serve)

You can change the properties of the material by adding any combination of the following calls:

- `.ingot()` will make the material have both an ingot and dust form.
- `.dust()` will make the material have a dust form. Don't use this together with `.ingot()`.
- `.gem()` will make the material have both a gem form and a dust form. Don't use those together with `.dust()` or `.ingot()`
- `.liquid()` will make the material have a liquid (fluid) form with liquid properties.

> - `.block()` will make the material have a placeable (block) fluid form. Requires .liquid().

Reviewer: ```suggestion
- `.block()` will make the material have a placeable (block) fluid form. Requires `.liquid()`.
```

- `.gas()` will make the material have a gas (fluid) form with gas properties.
- `.plasma()` will make the material have a plasma (fluid) form with plasma properties.
- `.polymer()` will make the material have a dust form with polymer properties.
- `.ore()` will create an ore from the material.
  - Optionally you can add any of these sets of parameters: 
    1. is emissive -> `true` for emissive textures
    2. ore multiplier and byproduct multiplier -> how many crushed ores will be given from one raw ore and how many byproducts dusts will be given throughout the ore processing 

>     3. ore multiplier, byproduct multiplier, is emissive

Reviewer: Needs more indent, isn't actually indented in mkdocs

Reviewer: <img width="863" height="245" alt="image" src="https://github.com/user-attachments/assets/0bd0ced7-6ee3-4b00-858f-2074d4bc3e32" />

- `.burnTime(int burnTime)` will turn the material into a furnace fuel.
- `.fluidBurnTime(int burnTime)` defines how long the fluid of the material will burn.
- `.components(component1, component2, ...)` describes the composition. The components are a list of elements of the following form: `'Kx material_name'`, where `K` is a positive integer.
- `.element(element)` -> similar to `.components()`, but is used when the material represents an element.
- `.iconSet(set)` gives the material an icon set.

> - `.color(int colorCode)` gives the material a color. The color must be provided as a hex value in the following form: `0xNNNNNN`, where `N` are digits.

Reviewer: ```suggestion
- `.color(int colorCode)` gives the material a color. The color must be provided as a hex value in the following form: `0xRRGGBB`.
```

- `.secondaryColor(int colorCode)` gives the material a secondary color. If this is not being called, the secondary value will default to white(0xffffff).

> - The secondary color is the overlay over the primary color on the material. This can be seen in the dust of a material, as the secondary color outline is visible. Rotors are another solid example.

Reviewer: This should be indented

Reviewer: ![Uploading image.png…]()

- `.addDefaultEnchant(string EnchantName, int level)` gives the material a default enchant. 

!!! tip "Harvest Level & Burn Time"
    For `.ingot()`, `.dust()` and `.gem()`, optionally you can put inside the parentheses any of these sets of parameters:

    1. harvest level (e.g. `.ingot(2)` will make the material have the harvest level of iron tools) 
    2. harvest level, burn time (e.g. `ingot(2, 2000)` will make the material have the harvest level of iron tools and will burn in furnaces as fuel for 2000 ticks or 100 seconds).

!!! tip "Choosing EU/t"
    GT has some builtin constants to ease choosing the required EU/t:

>     - `GTValues.V` for a full amp of power at the selected tier

Reviewer: This list isn't indented in the web view
<img width="809" height="128" alt="image" src="https://github.com/user-attachments/assets/f2395760-df41-485a-bd98-8e8bc0e1e62e" />

Materials are in-game items or fluids. They can be dusts, ingots, gems, fluids and all their derivatives.
To make a new material, write an `event.create()` call in the registering function, like in the examples.
Write inside the parentheses the name of the material inside `''` or `""`.

> (**NOTE**: to add a material that is present on the periodic table, but doesn't have any in-game items/fluids, look at the [material modification page](./Modifying-Existing-Materials.md))

Reviewer: This should probably get turned into a `!!! note`

You can change the properties of the material by adding any combination of the following calls:

- `.ingot()` will make the material have both an ingot and dust form.
- `.dust()` will make the material have a dust form. Don't use this together with `.ingot()`.

> - `.gem()` will make the material have both a gem form and a dust form. Don't use those together with `.dust()` or `.ingot()`

Reviewer: ```suggestion
- `.gem()` will make the material have both a gem form and a dust form. Don't use those together with `.dust()` or `.ingot()`.
```

- `.liquid()` will make the material have a liquid (fluid) form with liquid properties.
- `.block()` will make the material have a placeable (block) fluid form. Requires `.liquid()`.
- `.gas()` will make the material have a gas (fluid) form with gas properties.
- `.plasma()` will make the material have a plasma (fluid) form with plasma properties.
- `.polymer()` will make the material have a dust form with polymer properties.

> - `.ore()` will create an ore from the material.
>     - Optionally you can add any of these sets of parameters: 
>         1. is emissive -> `true` for emissive textures
>         2. ore multiplier and byproduct multiplier -> how many crushed ores will be given from one raw ore and how many byproducts dusts will be given throughout the ore processing 
>         3. ore multiplier, byproduct multiplier, is emissive

Reviewer: ```suggestion
- `.ore()` will create an ore from the material.
    - Optionally you can add any of these sets of parameters: 
        1. `boolean isEmissive` -> `true` for emissive textures
        2. `int oreMultiplier, int byproductMultiplier` -> how many crushed ores will be given from one raw ore and how many byproducts dusts will be given throughout the ore processing 
        3. `int oreMultiplier, int byproductMultiplier, boolean isEmissive` -> see previous points
```

- `.burnTime(int burnTime)` will turn the material into a furnace fuel.
- `.fluidBurnTime(int burnTime)` defines how long the fluid of the material will burn.
- `.components(component1, component2, ...)` describes the composition. The components are a list of elements of the following form: `'Kx material_name'`, where `K` is a positive integer.

> - `.element(element)` -> similar to `.components()`, but is used when the material represents an element.

Reviewer: ```suggestion
- `.element(element)` is similar to `.components()`, but is used when the material represents an element.
```
to keep consistency with the rest

> Materials are in-game items or fluids. They can be dusts, ingots, gems, fluids and all their derivatives.
> To make a new material, write an `event.create()` call in the registering function, like in the examples.
> Write inside the parentheses the name of the material inside `''` or `""`.

Reviewer: ```suggestion
Materials are in-game items or fluids. They can be dusts, ingots, gems, fluids and all their derivatives.
```

Reviewer: since we show how to do this properly in the examples now

    GTCEuStartupEvents.registry('gtceu:material', event => {
        event.create('mysterious_ooze')
          .fluid() // Or .liquid(Int Temperature)
          .color(0x500bbf)
    })
    ```

=== "Java"
    ```java title="fluid.java"
    public static Material MYSTERIOUS_OOZE;
    public static void register() {
        MYSTERIOUS_OOZE = new Material.Builder(
            your_mod_id.id('mysterious_ooze'))
            .fluid() // Or .liquid(Int Temperature)
            .color(0x500bbf)
            .buildAndRegister();
        }
    ```

!!! tip "Further Material Information"
    For more information on more fine grained material control, check out the pages below!

For a full list of the flags, check out the [material flags page](https://gregtechceu.github.io/GregTech-Modern/Modpacks/Materials-and-Elements/Material-Flags/).

For a full list of material properties, check out the [material properties page](https://gregtechceu.github.io/GregTech-Modern/Modpacks/Materials-and-Elements/Material-Properties/).

For an explanation of tools, check out the [tool creation page](https://gregtechceu.github.io/GregTech-Modern/Modpacks/Materials-and-Elements/Tool_Properties/).

> For an explanation of custom icon sets and a list of existing ones, check out the [icon set page](https://gregtechceu.github.io/GregTech-Modern/Modpacks/Materials-and-Elements/Material-Icon-Sets/).

Reviewer: Do relative links instead of absolute here

    public static void register() {
        ANDESITE_ALLOY = new Material.Builder(
                your_mod_id.id("andesite_alloy"))
                .ingot()
                .components("1x andesite", "1x iron")
                .color(0xFF0000).secondaryColor(0x840707).iconSet(GTMaterialIconSet.DULL)
                .flags(MaterialFlags.GENERATE_PLATE, MaterialFlags.GENERATE_GEAR, MaterialFlags.GENERATE_SMALL_GEAR)
                .buildAndRegister();
        }
    ```

## Creating a Dust

=== "JavaScript"
    ```js title="dust.js"
    GTCEuStartupEvents.registry('gtceu:material', event => {
        event.create('mysterious_dust')
            .dust() // The haverst level and burn time can be specified in the brackets. Example: `.dust(2, 4000)`
            .color(0x7D2DDB)
    })
    ```

=== "Java"
    ```java title="Dust.java"
    public static Material MYSTERIOUS_DUST;
    public static void register() {
        MYSTERIOUS_DUST = new Material.Builder(
            your_mod_id.id("mysterious_dust"))
            .dust() // The haverst level and burn time can be specified in the brackets. Example: `.dust(2, 4000)`
            .color(0x7D2DDB)
            .buildAndRegister();

>     }

Reviewer: Change haverst to harvest

??? tip "Color Pickers"
    To chose a color for your material, you can checkout https://www.w3schools.com/colors/colors_picker.asp
    After you select a color with the above tool, copy the 6 digits that follow the # under the color preview.

## Creating an Ingot

=== "JavaScript"
    ```js title="ingot.js"
    GTCEuStartupEvents.registry('gtceu:material', event => {
        event.create('andesite_alloy')
            .ingot()
            .components('1x andesite', '1x iron')
            .color(0x839689).iconSet(GTMaterialIconSet.DULL)
            .flags(GTMaterialFlags.GENERATE_PLATE, GTMaterialFlags.GENERATE_GEAR, GTMaterialFlags.GENERATE_SMALL_GEAR)
    })
    ```
=== "Java"
    ```java title="Ingot.java"
    public static Material ANDESITE_ALLOY;
    public static void register() {
        ANDESITE_ALLOY = new Material.Builder(
                your_mod_id.id("andesite_alloy"))
                .ingot()

>                 .components('1x andesite', '1x iron')

Reviewer: Java strings have to use " I think, also applies for all other java examples

            .ore(2, 3) 
            .color(0x7D2DDB).iconSet(GTMaterialIconSet.LIGNITE)
            .buildAndRegister();
        }
    ```

## Creating a Fluid

=== "JavaScript"
    ```js title="fluid.js"
    GTCEuStartupEvents.registry('gtceu:material', event => {
        event.create('mysterious_ooze')
          .fluid() // Or .liquid(Int Temperature)
          .color(0x500bbf)
    })
    ```

=== "Java"
    ```java title="Fluid.java"
    public static Material MYSTERIOUS_OOZE;
    public static void register() {
        MYSTERIOUS_OOZE = new Material.Builder(
            your_mod_id.id('mysterious_ooze'))
            .fluid() // Or .liquid(Int Temperature)
            .color(0x500bbf)
            .buildAndRegister();
        }
    ```

!!! tip "Further Material Information"
    For more information on more fine grained material control, check out the pages below!

For a full list of the flags, check out the [material flags page](./Material-Flags.md).

For a full list of material properties, check out the [material properties page](./Material-Properties.md).

For an explanation of tools, check out the [tool creation page](./Tool_Properties.md).

> For an explanation of custom icon sets and a list of existing ones, check out the [icon set page](./Material-Icon-Sets.md).

Reviewer: Capitalize the words in the clickable part of the links here so they match the page names as they are in the sidebar
